// Components/Cell.js
var Block = require('../Block')
var proto = require('proto')
var Style = require("Style")

module.exports = proto(Block, function(superclass) {

	// static properties

	this.name = 'TableCell'

    this.defaultStyle = Style({
        display: 'table-cell'
    })
	

	// instance properties

	this.init = function(/*[label,] contents*/) {
        if(arguments.length <= 1) {
            var contents = arguments[0]
        } else {
            var label = arguments[0]
            var contents = arguments[1]
        }

        this.domNode = document.createElement("td") // do this before calling the superclass constructor so that an extra useless domNode isn't created inside it
		superclass.init.call(this) // superclass constructor
		this.label = label

        if(contents instanceof Block || typeof(contents) !== 'string') {
            this.add(contents)
        } else if(contents !== undefined) {
            this.domNode.textContent = contents
        }
	}

	this.colspan = function(cols) {
		this.attr('colspan',cols);
	}
});
