// Components/Option.js
// note: this is  not intended to be used directly - only through Select and MultiSelect

var proto = require('proto')

var Gem = require('Gem')
var Style = require('Style')
var domUtils = require("domUtils")

// emits a 'change' event when its 'selected' value changes
module.exports = proto(Gem, function(superclass) {

    // staic members

    this.name = 'Option'

    this.defaultStyle = Style({
        display: 'block'
    })


    // instance members

    this.init = function(label, value, text) {
        this.domNode = document.createElement("option") // do this before calling the superclass constructor so that an extra useless domNode isn't created inside it
        superclass.init.call(this) // superclass constructor

        this.label = label

        this.text = text
        this.val = value
    }

    Object.defineProperty(this, 'val', {
        // returns the value of the Option
        get: function() {
            return this._value
        },

        // sets the value of the Option
        set: function(value) {
            if(this.parent !== undefined) {
                if(this.parent.options[value] !== undefined) {
                    throw new Error("Can't give an Option the same value as another in the Select or MultiSelect (value: "+JSON.stringify(value)+")")
                }

                if(this.val !== null) {
                    delete this.parent.options[this.val]
                }

                this.parent.options[value] = this
            }

            this._value = value

        }
    })


    Object.defineProperty(this, 'selected', {
        // returns whether or not the option is selected
        get: function() {
            return this.domNode.selected
        },

        // sets the selected state of the option to the passed value (true for selected)
        set: function(value) {
            var booleanValue = value === true
            if(this.selected === booleanValue) return false; // ignore if there's no change

            if(this.parent !== undefined)
                this.parent.prepareForValueChange([this.val])

            this.setSelectedQuiet(booleanValue)

            if(this.parent !== undefined)
                this.parent.emit('change')
        }
    })

    Object.defineProperty(this, 'text', {
        get: function() {
            return this.domNode[domUtils.textProperty]
        },

        set: function(text) {
            this.domNode[domUtils.textProperty] = text
        }
    })


    // private

    // does everything for setting the selected state except emit the parent's change event
    this.setSelectedQuiet = function setOptionSelected(booleanValue) {
        if(this.selected === booleanValue) return; // ignore if there's no change

        this.domNode.selected = booleanValue
        this.emit('change') // the browser has no listenable event that is triggered on change of the 'checked' property
    }
})
