// Components/Text.js
var proto = require('proto')

var Gem = require('Gem')
var Style = require("Style")

var domUtils = require('domUtils')



module.exports = proto(Gem, function(superclass) {

    //static properties

    this.name = 'Text'

    this.defaultStyle = Style({
        whiteSpace: 'pre-wrap' // so whitespace is displayed (e.g. multiple spaces don't collapse)
    })

    this.init = function(/*[label,] text*/) {
        if(arguments.length === 1) {
            var text = arguments[0]
        } else {
            var label = arguments[0]
            var text = arguments[1]
        }

        if (text === undefined) text = ''

        this.domNode = domUtils.div() // need the domNode before setting the label

        this.label = label
        this.text = text

        superclass.init.apply(this, arguments) // superclass constructor
    }

    // instance properties

    Object.defineProperty(this, 'text', {
        get: function() {
            return this.domNode[domUtils.textProperty]
        }, set: function(v) {
             this.domNode[domUtils.textProperty] = v
        }
    })
});
